Fix #34304: Document strictly constraint precedence behavior (#35673)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/reference/dependency-management/declaring-dependencies/dependency_constraints.adoc

include::sample[dir="snippets/dependencyManagement/managingTransitiveDependencies-versionsWithConstraints/groovy",files="build.gradle[tags=dependency-strictly]"]
====

[[sec:strictly-constraint-precedence]]
== Behavior of `strictly` Constraints at Different Dependency Graph Depths

> When multiple `strictly` version constraints are declared for the same dependency at different levels of the dependency graph, Gradle honors the constraint that is *closest to the root* of the graph and ignores constraints further from the root.

Reviewer: ```suggestion
When multiple `strictly` version constraints are declared for the same dependency at different points in the dependency graph, Gradle always selects the constraint that is *closest to the root* of the graph.
Constraints declared deeper in the graph, such as in transitive dependencies or subprojects, cannot override a parent project’s `strictly` constraint.
```

> This means that a `strictly` constraint declared in your project's direct dependencies will take precedence over `strictly` constraints declared in transitive dependencies or subprojects.

Reviewer: ```suggestion
In other words, a `strictly` constraint declared directly in your project’s dependency declarations takes precedence over any `strictly` constraints declared by the components you depend on.
```

.Root-level `strictly` constraint takes precedence
====

> In this example, the root project declares a `strictly` constraint for version 4.4, while a subproject declares a different `strictly` constraint for version 4.5.

Reviewer: ```suggestion
In this example, the root project declares a `strictly` constraint on version 4.4, while a subproject declares a `strictly` constraint on version 4.5:
```

The root-level constraint (4.4) takes precedence.

> [tabs]

Reviewer: This needs to be turned into a tested snippet, you can see how to do that there: https://github.com/gradle/gradle/tree/master/platforms/documentation#code-snippets

There are many examples you can use. I recommend copy pasting an existing one and adjusting the build files to include the code below.

    api(project(":moduleA"))
    api("org.apache.httpcomponents:httpclient") {
        version {
            strictly("4.4")  // This constraint is honored
        }
    }
}

// moduleA/build.gradle.kts
plugins {
    `java-library`
}

dependencies {
    constraints {
        api("org.apache.httpcomponents:httpclient") {
            version {
                strictly("4.5")  // This constraint is ignored
            }
        }
    }
}
----

Groovy::
+
[source,groovy]
----
// Root project build.gradle
plugins {
    id 'java-library'
}

dependencies {
    api(project(':moduleA'))
    api('org.apache.httpcomponents:httpclient') {
        version {
            strictly '4.4'  // This constraint is honored
        }
    }
}

// moduleA/build.gradle
plugins {
    id 'java-library'
}

dependencies {
    constraints {
        api('org.apache.httpcomponents:httpclient') {
            version {
                strictly '4.5'  // This constraint is ignored
            }
        }
    }
}
----
=====

In this scenario, version *4.4* of `httpclient` will be selected because the constraint in the root project is closer to the root of the dependency graph than the constraint in `:moduleA`.
====

> This behavior ensures that projects have control over their direct dependency versions and can override constraints from transitive dependencies.

Reviewer: ```suggestion
This behavior ensures that projects maintain control over the versions of their direct dependencies and can override any constraints introduced by transitive dependencies.
```
